mono/mini/mini-windows-tls-callback.c

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright 2019 Microsoft Corporation
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+#include "mini-runtime.h"
+
+#if defined(HOST_WIN32)
+#include "mini-windows.h"
+#include <windows.h>
+
+/* Only reset by initial callback from OS loader with reason DLL_PROCESS_ATTACH */
+/* both for DLL and LIB callbacks, so no need to protect this variable. */
+static MonoWin32TLSCallbackType mono_win32_tls_callback_type = MONO_WIN32_TLS_CALLBACK_TYPE_NONE;
+
+/* NOTE, this function needs to be in this source file to make sure linker */
+/* resolve this symbol from mini-windows.c, picking up callback and */
+/* included it in TLS Directory PE header. Function makes sure we only activate */
+/* one of the supported mono callback types at runtime (if multiple one has been used) */
+gboolean
+mono_win32_handle_tls_callback_type (MonoWin32TLSCallbackType callback_type)
+{
+	if (mono_win32_tls_callback_type == MONO_WIN32_TLS_CALLBACK_TYPE_NONE)
+		mono_win32_tls_callback_type = callback_type;
+
+	if (callback_type != mono_win32_tls_callback_type)
+		return FALSE;
+
+	return TRUE;
+}
+
+MONO_EXTERN_C
+VOID NTAPI mono_win32_tls_callback (PVOID module_handle, DWORD reason, PVOID reserved);
+
+MONO_EXTERN_C
```

Comment:
Why `extern "C"`?

Reply:
Was just kept inline with DllMain, but since it's just used within this source, it could do without it.